### PR TITLE
Crimson Isle Profile Viewer Page

### PR DESCRIPTION
**Pull request by Vixid1**
Added a profile viewer page for Crimson Isle data.

image

Added:

- Kuudra Run Count (Burning, Fiery and Infernal have 0's as placeholders for now until they are released)
- Dojo Test Scores
- Dojo Points and Rank
- Current Selected Faction and each of the reputations
- Last Matriarch Attempt (It was in the api so why not)

The last matriarch attempt timestamp should account for timezones since it takes it from the users JVM. (I believe)

### CODE REVIEW

**Comment by Mqisty** [comment]
ik it makes sense to have the black rank be black but maybe dont, its hard to read

**Comment by Vixid1** [comment]
> ik it makes sense to have the black rank be black but maybe dont, its hard to read

Yeah I guess, I've changed it to dark gray in a new commit so it's readable now

image

**Comment by NopoTheGamer** [comment]
Two things
1. https://github.com/NotEnoughUpdates/NotEnoughUpdates/blob/master/src/main/java/io/github/moulberry/notenoughupdates/NotEnoughUpdates.java#L262
 you need to add this for your pv tab
2. Crashed while /pv throwpo, I assume due to the fact I've never done crimson stuff https://hst.sh/raw/yuhuzoyeqi

**Comment by Vixid1** [comment]
Shows placeholder values for when there is missing data in the player's API.

image

File: src/main/java/io/github/moulberry/notenoughupdates/profileviewer/CrimsonIslePage.java [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+public class CrimsonIslePage extends GuiProfileViewerPage {
+
+		private static final ResourceLocation CRIMSON_ISLE = new ResourceLocation("notenoughupdates:pv_crimson_isle_page.png");
+
+		public static final ItemStack[] KUUDRA_KEYS = {
+			NotEnoughUpdates.INSTANCE.manager.jsonToStack(
```
Comment (romangraef) [inline src/main/java/io/github/moulberry/notenoughupdates/profileviewer/CrimsonIslePage.java:45]: please dont handle jsons directly and try to use apis like itemresolutionquery instead when possible

**Review by romangraef — DISMISSED** [review]

**Comment by hannibal002** [comment]
@Vixid1 merge needed

**Review by NopoTheGamer — APPROVED** [review]

**Comment by hannibal002** [comment]
Maybe try to integrate this page with the trophy fishing page. The information is useful, jut not so useful that it would qualify for a separate PV page.

File: src/main/java/io/github/moulberry/notenoughupdates/profileviewer/CrimsonIslePage.java [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+				guiTop + 30,
+				130
+			);
+
+			int mageReputation = data.has("mages_reputation") ? data.get("mages_reputation").getAsInt() : 0;
+			int barbarianReputation = data.has("barbarians_reputation") ? data.get("barbarians_reputation").getAsInt() : 0;
+
+			LinkedHashMap<Integer, String> factionThresholds = new LinkedHashMap<Integer, String>() {{
+				put(0, "Neutral");
+				put(1000, "Friendly");
```
[the remark this answers is not in the document]
Comment (Vixid1) [reply src/main/java/io/github/moulberry/notenoughupdates/profileviewer/CrimsonIslePage.java:346]: I believe so, it wasn't in the wiki when I checked for the exact thresholds so I forgot about them. I've added them now though.

**Comment by hannibal002** [comment]
All lines in the CrimsonIslePage class (and maybe other classes too?) are wrongly formatted, according to my IntelliJ auto formatter (control+alt+L).

Perhaps merge drawMageReputation and drawBarbarianReputation into one function.

Most of the static data in the maps can be moved into the repo, this has very low priority.

**Review by hannibal002 — APPROVED** [review]
Looking good to me.

About `CrimsonIslePage.java`: I would prefer a class with ~300 lines, but without all the expanded `renderAlignedString` functions the size is at 310 (instead of the current 404 lines), so it's fine.

**Review by NopoTheGamer — APPROVED** [review]

**Review by Lulonaut — APPROVED** [review]

**Pull request merged**